internal/controller/bucket_controller.go

```diff
@@ -0,0 +1,225 @@
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/giantswarm/object-storage-operator/api/v1alpha1"
+	managementcluster "github.com/giantswarm/object-storage-operator/internal/pkg/managementcluster"
+	"github.com/giantswarm/object-storage-operator/internal/pkg/service/objectstorage"
+)
+
+// BucketReconciler reconciles a Bucket object
+type BucketReconciler struct {
+	client.Client
+	objectstorage.ObjectStorageServiceFactory
+	managementcluster.ManagementCluster
+}
+
+//+kubebuilder:rbac:groups=objectstorage.giantswarm.io,resources=buckets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=objectstorage.giantswarm.io,resources=buckets/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=objectstorage.giantswarm.io,resources=buckets/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the bucket closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.16.0/pkg/reconcile
+func (r BucketReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	logger.Info("Started reconciling Bucket")
+	defer logger.Info("Finished reconciling Bucket")
+
+	// Get the bucket that we are reconciling
+	bucket := &v1alpha1.Bucket{}
+	err := r.Client.Get(ctx, req.NamespacedName, bucket)
+	if err != nil {
+		return ctrl.Result{}, errors.WithStack(client.IgnoreNotFound(err))
+	}
+
+	// Create the correct service implementation based on the provider
+	var service objectstorage.ObjectStorageService
+	switch r.ManagementCluster.Provider {
+	case "capa":
+		roleArn, err := r.getRoleArn(ctx)
+		if err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+
+		service, err = r.ObjectStorageServiceFactory.NewS3Service(ctx, r.ManagementCluster.Region, roleArn)
+		if err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+	default:
+		return ctrl.Result{}, errors.New(fmt.Sprintf("unsupported provider %s", r.ManagementCluster.Provider))
+	}
```

Comment:
It cannot be done in main because getting the role to assume should be done in every reconciliation loop. Yes @fiunchinho told me the same thing but this looks easier to me (and I can add unknown provider tests already)